Draft Explainer (#309)
w3c/wot-discovery: explainer/Explainer.md

It is an optional feature of the WoT Architecture in that WoT TDs can also be 
obtained by other means, for example by using a normal database.  However, 
WoT Discovery is meant to address use cases where IoT devices from different
vendors need to be integrated, or where a large number of IoT services need to be 
searched in a consistent way.

## Introduction

## Relationship to Other WoT Deliverables
For a general introduction to the Web of Things (WoT) Architecture, including a description
of all normative and informative deliverable, please see the 
[WoT Architecture explainer](https://github.com/w3c/wot-architecture/blob/main/explainer/explainer11.md)
and the [WoT Architecture document](https://w3c.github.io/wot-architecture/).

Since the purpose of the WoT Discovery deliverable is to normatively
define a mechanism to distribute WoT Thing Descriptions,
the [WoT Thing Description 1.1 document](https://w3c.github.io/wot-thing-description/) is
especially relevant.  The WoT Discovery specification also includes a Thing Model,
defined in this deliverable,
which normatively defines many aspects of the Thing Description Directory service specified

> as part of the discovery process.  

Author: going to stick with top-down

## Use Cases and Requirements
In general, WoT Use Case and Requirements have been collected from a large number 
of participants and are collected in the 
[WoT Use Cases and Requirements document](https://w3c.github.io/wot-usecases/).
In the following we extract a few of the most critical requirements and motivating
use cases driving the design of WoT Discovery.

### Integration
It is often the case in IoT systems that devices provided by different manufacturers

> need to be integrated together.  For example, consider a Retail use case

Author: fixed

services, probably selected from a large number, that meet some criteria, such
as location or type.

### Privacy and Security
While the purpose of the WoT Discovery process is to distribute Thing Description metadata,
it must also satisfy security and privacy constraints.
In general, TDs should only be distributed to entities after their identities have been
confirmed (authentication) and it has been established that they have the rights to view that data
(authorization).  In addition, data needs to be protected from interception and modification,
and it is also necessary to guard against other risks such as tracking or inferencing of
private data.  

The WoT Discovery architecture has been defined with safeguarding security and privacy 
as a primary goal.

The WoT Architecture and WoT Thing Description documents have additional security and
privacy considerations that are relevant to WoT Discovery.  In particular, 
the WoT Architecture document discusses when and where secure transport such as TLS
should be used, and when various access control mechanisms should be used.
The problem is complicated by the fact that we often want to access IoT devices

> on a LAN where the CA system available on the rest of the internet is not easy

Reviewer: I think some abbreviations like CA could/should be spelled out the first time they are being used (although the explainer's target audience is probably familiar with the terminology).

Author: yeah, I was thinking that too.  Expanded.

to apply.  In general we require strong controls on the distribution of data from
WoT Things in general and WoT Discovery results in particular when access is public, and 
best-effort controls on private networks such as LANs.

## Architecture
The architecture has been designed with the following goals in mind:
- Security and privacy.
- Ease of use.
- Integration with existing ecosystems.
- Both local and global access.
- Scability to large numbers of Things.

> - Support for both self-description and external descriptions.

Author: tweaked to use initial lower case and semicolons to make the whole thing one big run-on sentence.  An improvement?

We use a two-phase approach.  In the first phase, which we call "Introductions",
we support any number of open discovery mechanisms whose results, however,
are constrained to simple URLs.  These URLs should not contain any metadata themselves.
Instead, they have to be resolved into "Exploration" services which can be used
to retreive metadata. Exploration services, however, are tasked with 
authenticating and checking the authorizations of requests, and protecting
queries and the transmission of results, and can use best
practices from web services to do so, including (for example) OAuth and TLS.

### Introduction Mechanisms
Introduction mechanisms allow for ecosystem integration and using different "first contact"
mechanisms in different circumstances.  For example, in a Smart City, we might
want to use a normal web search to find suitable WoT Discovery exploration
services for that city.  In a Smart Home, we might use other mechanisms 
restricted to a single LAN, such as mDNS in combination with DNS-SD.  If simply
want to tell someone where to go, we can just use written URLs or QR codes.

The security and privacy of WoT Discovery does not depends on Introduction

> mechanisms being secure.  They must ONLY distribute URLs, and those URLs should

Author: sure

not have metadata embedded in them.

### Exploration Mechanisms
Exploration mechanisms come in two different forms: directories and self-description.
Both of these are HTTP-based web services and support existing mechanisms

> for authentication, authorization, and secure transport.

Author: https://github.com/w3c/wot-discovery/issues/314

> for authentication, authorization, and secure transport.

Author: proposed resolution: clean up spec for review, HTTP-only; IF we can get a PR to support CoAP for self-description at least, then we can extend this slightly, but we are running out of time, and the spec as it is is HTTP-only, pretty much.

A WoT Thing Description Directory (TDD) is an HTTP-based Thing itself, described
with a Thing model.
A TDD supports
a queryable interface to a database of Thing Descriptions.  This supports
use cases like a Smart City, Factory, or even a Smart Home where there
might be hundreds or thousands of Things to select from.

Self-description is supported by the ability of an HTTP-based Thing to

> return a TD directly.  There is also an Introduction mechanism that allows

Author: as above...

devices built from the outset to support the Web of Things.  For "brownfield"
devices the TD describing how to access a devices must be obtained elsewhere.
In this case a TDD is a better solution also.

## Non-Goals
- Onboarding: WoT Discovery and WoT in general only define the behavior of devices
after they have been provisioned (provided with identities and associated keys and
certificates) and access controls for users have been allocated.  In other words,
WoT Discovery is designed to validate existing credentials, not to establish them.
- Distribution of Security Data: WoT Discovery, and WoT Thing Descriptions,
are not designed for the distribution of security information, including public
keys.  The latter are technically public but additional controls are needed to ensure
their integrity, and there are existing mechanisms to do so (e.g. certificates)
that should be used.  WoT Discovery and WoT TDs, however, do contain security *metadata*
about how a Thing can be accessed, *if* the requestor has access to the associated
private security data (tokens, passwords, etc).
- Definition of New Security Mechanisms: WoT Discovery uses existing security

> mechanisms to provide the necessary protection.  It does not define new ones.

Author: yeah, making these subsections would be a good idea.  I also just added a number of new ones.

cannot be modified and so can only be described externally.

The WoT Discovery specification attempts to meet these challenges by defining
a multistage discovery process with authorization checkpoints to meet privacy and
security goals; by using scalable RESTful API designs; by supporting a variety of
first-contact mechanisms suitable for different scales; and by providing optional
semantic search and self-description capabilities.

There are still a number of limitation in the current specification, which
primarily focuses on HTTP and had to defer some important goals, such as
geolocation queries.  However, the specification attempts to address the above
core issues and is intended to serve as a foundation for future work while
enabling, in its current form, use cases supporting IoT integration
and access to public IoT services.

## Relationship to Other WoT Deliverables
For a general introduction to the Web of Things (WoT) Architecture, including a description
of all normative and informative deliverables, please see the 
[WoT Architecture explainer](https://github.com/w3c/wot-architecture/blob/main/explainer/explainer11.md)

> and the [WoT Architecture document](https://w3c.github.io/wot-architecture/).

Reviewer 2: Are github.io URLs stable and valid in long term?
```suggestion
and the [WoT Architecture document](https://www.w3.org/TR/wot-architecture11/).
```

Since the purpose of the WoT Discovery deliverable is to normatively
define a mechanism to distribute WoT Thing Descriptions,

> the [WoT Thing Description 1.1 document](https://w3c.github.io/wot-thing-description/) is

Reviewer 2: Alt text mentions versioned as 1.1
```suggestion
the [WoT Thing Description 1.1 document](https://www.w3.org/TR/wot-thing-description11/) is
```

especially relevant.
The WoT Discovery specification also includes a Thing Model,
which normatively defines many aspects of the Thing Description Directory service specified
as part of the discovery process.
Thing Models, as defined in the Thing Description 1.1 document,
are a format for describing classes of Things.
We model Thing Description Directories as a special kind of Thing.

## Use Cases and Requirements
In general, WoT Use Cases and Requirements have been collected from a large number 
of participants and are collected in the 

> [WoT Use Cases and Requirements document](https://w3c.github.io/wot-usecases/).

Reviewer 2: ```suggestion
[WoT Use Cases and Requirements document](https://www.w3.org/wot-usecases/).
```
https://www.w3.org/wot-usecases/ is not found for some reason!

Reviewer 3: "https://www.w3.org/TR/wot-usecases/" ?

### Use Case: Public IoT Services
In a Smart City, we may want to provide access to IoT devices as a public service,
for example to report local air quality conditions.  In Retail, we may want to
make available semi-public services (perhaps requiring pre-registration) to access
accessiblity devices, such as chair lifts.  In this case, a user wants to discover
services, probably selected from a large number, that meet some criteria, such
as location or type.

### Requirements
WoT Discovery has been designed with the following goals in mind:
- security and privacy;
- ease of use;
- integration with existing ecosystems;
- both local and global access;
- scalability to large numbers of Things, including distributing searches over multiple directory services; and
- support for both self-description and external descriptions (to support brownfield scenarios).

Some of these are in conflict.  For example, while we want to integrate with
existing ecosystems, we don't want to expose users to known or potential 
security hazards in those ecosystems.
Making discovery simple enough to be used on small devices for self-description
while also secure and

> scalable to a large number of IoT devices is also a significant challenge.

Reviewer 4: It is a mild repetition of what has been said in the introduction, but it might not hurt too much.

simple enough to implement on individual IoT devices but with good potential for scalability for
large cloud-based systems.

## Architecture
In general, WoT Discovery is about distribution of TDs, while
privacy and security are about restricting distribution of information
only to verified authorized parties.  Taken together, WoT Discovery should enable
the distribution of TDs only to verified and authorized parties.

The WoT Discovery architecture is designed to address these challenges.
We use a two-phase approach.  In the first "first-contact" phase, which we call "Introductions",
we support any number of open discovery mechanisms whose results, however,
are constrained to simple URLs.  These URLs should not contain any metadata themselves.
Instead, they have to be resolved into detailed-search "Exploration" services which can be used
to retreive metadata. Exploration services, however, are tasked with 
authenticating and checking the authorizations of requests, and protecting
queries and the transmission of results, and can use best
practices from web services to do so, including (for example) OAuth and TLS.

> Exploration services implement a scalable RESTful web API.

Reviewer 4: I agree, what about saying: Exploration services might implement a scalable RESTful web API.? However, if the explainer will live and die with the current status of the spec document. It might be ok to assume Exploration like TDDs or RESTFul we API.

> Exploration services implement a scalable RESTful web API.

Author: Good point!   I have pushed an update tweaking this paragraph and explicitly mentioning both directories and self-description.

### Introduction Mechanisms
Introduction mechanisms allow for ecosystem integration and using different "first contact"
mechanisms in different circumstances.  For example, in a Smart City, we might
want to use a normal web search to find suitable WoT Discovery exploration
services for that city.  In a Smart Home, we might use other mechanisms 
restricted to a single LAN, such as mDNS in combination with DNS-SD.  If we simply
want to tell someone where to go, we can just use written URLs or QR codes.

The overall security and privacy of WoT Discovery does not depend on Introduction
mechanisms being secure.  Introductions must *only* distribute URLs, and those URLs should

> not have metadata embedded in them.  Additional authentication and authorization

Author: hmm.  changed to "Other than some optional and very generic information about exploration service type,
introductions must *only* distribute URLs,"

Makes it more verbose, but oh, well.

is still needed before detailed metadata is provided.

### Exploration Mechanisms
Exploration mechanisms come in two different forms: directories and self-description.
Both of these are HTTP-based RESTful web services and support existing mechanisms
for authentication, authorization, and secure transport.

A WoT Thing Description Directory (TDD) is an HTTP-based Thing itself.
As a class they are described with a Thing Model and any particular
instance also has a fetchable Thing Description.
A TDD supports
a queryable interface to a searchable database of Thing Descriptions.
These Thing Descriptions can describe IoT devices or other directories,
either directly or by reference (Thing Links).
This supports
use cases like a Smart City where there
might be thousands of Things to select from.
However, it can also be used in a Smart Home with only a few dozen Things.

Self-description is another exploration mechanism that
provides the ability of an HTTP-based Thing to
return a TD directly.  There is also an Introduction mechanism that allows
the URL for self-description on a Thing to be guessed using a well-known

> URL convention based only on the base IP address of the Thing.  In this
> exploration mechanism, however, the Thing itself is responsible for managing
> authentication and authorization, so it is only suitable for a small number
> of users and for managing a small number of devices.  This is also an optional
> feature.  It is also worth noting that it only applies to "greenfield"

Reviewer 2: ```suggestion
URL convention based only on the base IP address of the Thing.  In this
exploration mechanism, however, the Thing itself is responsible for managing
authentication and authorization, so it is only suitable for a small number
of users and for managing a small number of devices.  This is an optional
feature.  It is also worth noting that it only applies to "greenfield"
```

about how a Thing can be accessed, *if* the requestor has access to the associated
private security data (tokens, passwords, etc).

### Definition of New Security Mechanisms
WoT Discovery uses existing security
mechanisms to provide the necessary protection.  It does not define new ones.

### Discovering Thing Models
WoT Discovery currently supports the distribution
of Thing Descriptions, which describe actual instances of Things. Thing Models
are a recent addition to the WoT architecture which permit the description of 
parameterized classes of Things. 

Technically, WoT Discovery could be extended to also support searching
for Thing Models (which are also JSON-LD) but we have deferred this 
possible support to future work
until we better understand the use cases.

### Historical Data
WoT Discovery focuses on searching metadata (Thing Descriptions),
not searching actual data or historical data returned from IoT devices.

## Design Decisions and Limitations

### No Required Query Language
We considered three different query languages: JSONPath, XPath, and SPARQL.
Unfortunately XPath, while it was recently extended to support JSON, does

> not yet have enough suitable implementations of version 1.3 supporting this extension.

Reviewer 2: ```suggestion
not yet have enough suitable implementations of version 3.1 supporting this extension.
```

SPARQL is suitable for semantic search but too heavy a requirement for small IoT hubs to
make mandatory.
We would have liked to make JSONPath mandatory as it is good mix of functionality
and lightness but it is not yet standardized.  An RFC is under discussion within
IETF but the current draft JSONPath draft proposal is missing some key features we need.

In the long run (in a future specification cycle) we would like to add a standardized
version of JSONPath as a required query language so for now we have only specified an optional 

> interface for SPARQL 1.1.  Interfaces for XPath 1.3 and JSONPath

Reviewer 2: ```suggestion
interface for SPARQL 1.1.  Interfaces for XPath 3.1 and JSONPath
```

There are also use cases for small directories with static content (e.g. on
IoT devices co-hosting a small, fixed set of Things) where a query language would
be difficult to justify.

### IDs
The Thing Description specification includes optional IDs which are mapped onto
the RDF @id.  However, to support SPARQL queries, an @id is necessary in the 
information model.  Therefore
when an "anonymous" TD without an ID is registered with a Thing Description Directory,
a temporary ID is created that is local to the directory.

### Everything is a Thing
Directories maintain lists of Thing Descriptions but are also Things 
with their own Thing Description.

The reason for this is the two-phase introduction/exploration process.
There are introduction mechanisms that can only return a URL but cannot
associate a "type" with that URL. A client would then not know whether the URL is pointing
at a Thing Description or a Thing Description Directory.  We resolved this by deciding that
URLs returned by introduction mechanisms always point at Thing Descriptions,
but Thing Descriptions for Thing Description Directories have a special @type value.

### Static Directories

> The only required affordance in a Thing Description Directory is "things",
> which lists all the contents of the directory (with support for pagination, however,
> to deal with long responses).  The ability to register, update, and delete

Reviewer 2: ```suggestion
The only required affordances in a Thing Description Directory are "retrieveThing" and "things",
which query one or all the entries in the directory (with support for incremental transfer and pagination, however,
to deal with long responses).  The ability to register, update, and delete
```

Thing Descriptions in a directory is optional.

The reason for this is to support the case
of a small device that hosts multiple Things. For example, there exist
"multisensors" that host multiple independent sensors for temperature, humidity,
illumination, vibration, etc.  It is convenient for automation rules if
each of these is modelled as a separate Thing. However, for self-description,
the introduction URL needs to point at a single Thing Description.  We considered
adding a Thing Descriptions with links representing items in a collection, 
but settled on allowing small Thing Description Directories with static contents
for this use case. 

### Thing Links
We wanted to support "federated" directory services where a query can be
distributed over a set of directories.  SPARQL supports this feature but
requires the set of federated directories to be given explictly in a query.
For information modelling reasons, we wanted directories to also hold only
a set of Thing Descriptions.  In order for directories to refer to other 
directories for the purposes of federation, we support a special type of 
Thing Description, a Thing Link, whose purpose is to refer to another

> Thing Description at another location.  This includes Thing Descriptions of

Reviewer 2: I think directory federation was never the intention for introducing [Thing Links](https://w3c.github.io/wot-discovery/#exploration-td-type-thinglink). It rather to allow referencing large TDs (from a small device to a powerful server) and dynamic TDs (from a directory to a device at the edge). 
We recently added statements in [Architecture](https://w3c.github.io/wot-discovery/#architecture) and [Discovery Process](https://w3c.github.io/wot-discovery/#discoverer-process) suggesting the use of Think Link for linking directories.

Author: pls fix-up later

directories, which are a kind of Thing.  Then federated searches can be set
up by following these Thing Links.  We did not, however, make recursive searches
automatic on the server side to avoid accidental or malicious exponential
resource consumption.  Instead the client needs to explictly request searches
from linked directories, and also has to have access rights to such linked

> directories.

Reviewer 4: Might be useful to cite web crawlers here? the behavior explained above looks a lot like what a web crawler would do.

Author: might be helpful, suggest a change with a PR if you'd like.

### No Standardized Geolocation Data or Geolocation Search
Supporting geolocation was originally in our charter but unfortunately we had to defer
it.  We do have a draft proposal but need to do more work to align with existing
geolocation standards and consider privacy and security implications.  In addition,
introduction mechanisms and standardized query languages 
supporting geolocation search are currently lacking.

However, supporting this functionality is still a long-term goal for future work,
as it is important in many of the WoT use cases we have collected.
IEEE P2874 is developing standardized spatial queries and the
Open Geospatial Consortium is developing RDF information models for geolocation
data; both of these may be relevant to future work.

## Implementations

> Implementations are described in more detail in the WoT Discovery Implementation Report,
> but in summary the following four implementations of the WoT Discovery TDD service
> has been performed:
> - LogiLabs TDD - An implementation supporting SPARQL queries, sponsored by Siemens.
> - WoT Hive TDD - An implementation supporting SPARQL queries, sponsored by Universidad Politécnica de Madrid.
> - LinkSmart TDD - An early implementation done by Fraunhofer.
> - TinyIoT TDD - An implementation supporting JSONPath queries, forked from an 
>   the early implementation done by Fraunhofer.
> - Minimal TDD - An implementation that intentionally includes only the core
>   features designed to support static implementations, such as devices statically
>   hosting multiple separate sensors.  Done by Intel.

Reviewer 2: ```suggestion
Implementations are described in more detail in the WoT Discovery Implementation Report,
but in summary the following implementations of the WoT Discovery TDD service
has been performed:
- LogiLabs TDD - An implementation supporting SPARQL queries, sponsored by Siemens.
- WoT Hive TDD - An implementation supporting SPARQL queries, sponsored by Universidad Politécnica de Madrid.
- LinkSmart TDD - An early implementation done by Fraunhofer.
- TinyIoT TDD - An implementation supporting JSONPath queries, forked from an 
  the early implementation done by Fraunhofer.
- Minimal TDD - An implementation that intentionally includes only the core
  features designed to support static implementations, such as devices statically
  hosting multiple separate sensors.  Done by Intel.
```

## Implementations
Implementations are described in more detail in the WoT Discovery Implementation Report,
but in summary the following four implementations of the WoT Discovery TDD service
has been performed:
- LogiLabs TDD - An implementation supporting SPARQL queries, sponsored by Siemens.
- WoT Hive TDD - An implementation supporting SPARQL queries, sponsored by Universidad Politécnica de Madrid.
- TinyIoT TDD - An implementation supporting JSONPath queries, forked from an 
  initial implmentation done by Fraunhofer as part of the LinkSmart project.
- Minimal TDD - An implementation that intentionally includes only the core
  features designed to support static implementations, such as devices statically
  hosting multiple separate sensors.  Done by Intel.
Except for the last implementation, all of these WoT TDD implementations also
support registration and management of a dynamic database of TDs.

There are other aspects of WoT Discovery that also need to be considered.
We define a "Discoverer" as an entity that uses the WoT Discovery process to retrieve
a TD.  We have a number of implementations of Discoverers, including integration into the 

> Eclipse node-wot project (an implementation of the informative 

Reviewer 5: FYI: node-wot, at the moment, does not yet support discovery.
I think @JKRhb started work on discovery within his Dart implementation.

Author: Well, even just giving a URL pointing at a TD is considered "Discovery" - it's the "Direct" method.  Talked to Jan, and so far that's all he does (called "fetch").   However, security is a problem due to a lack of standard for setting up credentials, etc.  Also mapping credentials to id does not work when doing discovery, since you don't have the id yet.  This is also a security bootstrapping issue - which implementations like Dart and node-wot (and the Scripting API) need to be structured to support.

However... in the meantime, I will take node-wot out and add Dart, and Jan and I will talk about how to best support at least some aspects of Discovery in Dart in the short term.

Reviewer: An example for the discovery implementation in `dart_wot` can be seen here: https://github.com/namib-project/dart_wot/blob/main/example/complex_example.dart#L129-L136 It is currently quite basic and does not support security schemes, but it reflects the current state of the Scripting API when it comes to discovery (at least for the direct method, the directory method is not supported yet).

Regarding node-wot, I meant that there is at least a `fetch` helper function defined (https://github.com/eclipse/thingweb.node-wot/blob/master/packages/core/src/helpers.ts#L152), which sort of acts like the direct method. So I think you could say that node-wot also qualifies as a discoverer. Not sure if this is enough to keep it in the explainer, though?

Reviewer 5: I see Jans point... `direct` is supported in node-wot.
